rtos/MessagePipe.h

```diff
@@ -0,0 +1,264 @@
+ A message is a memory block that is send to a thread or interrupt service routine.
+ The MessagePipe should be configured to be not blocking when used in interrupt service routine to prevent blocking calls.
+
+  @tparam  T         data type of a single message element.
+  @tparam  queue_sz  maximum number of messages in pipe.
+
+*/
+
+template<typename T, uint32_t pipe_sz>
+class MessagePipe : public mbed::FileHandle, private mbed::NonCopyable<MessagePipe<T, pipe_sz> > {
+
+public:
+    MessagePipe()
+    {
+        set_blocking(true);
+    }
+
+    /** Equivalent to POSIX poll(). Derived from FileHandle.
+     *  Provides a mechanism to multiplex input/output over a set of file handles.
+     */
+    virtual short poll(short events) const {
+        short revents = 0;
+
+        if (!_mail.empty() && (events & POLLIN)) {
+            revents |= POLLIN;
+        }
+
+        if (!_mail.full() && (events & POLLOUT)) {
+            revents |= POLLOUT;
+        }
+
+        return revents;
+    }
+
+    /** Write the contents of a buffer to a file
+     *
+     *  @param buffer   The buffer to write from
+     *  @param length   The number of bytes to write
+     *  @return         The number of bytes written, negative error on failure
+     */
+    virtual ssize_t write(const void* buffer, size_t length) {
+        
+        if(length != sizeof(T)) {
+            return -EINVAL;
+        }
+
+        if(_mail.full() && !_blocking) {
+            return -EAGAIN;
+        }
+
+        T * message = _mail.alloc(osWaitForever);
+
+        if(message == NULL) {
+            return -EIO;
+        }
+
+        memcpy(message, buffer, sizeof(T));
+
+        bool mail_was_empty = _mail.empty();
+
+        _mail.put(message);
+
+        //notify on state change 
+        if(_mail.full() || mail_was_empty) {
+            wake();
+        }
+        
+        return sizeof(T);
+    }
+
+    /** Read the contents of a file into a buffer
+     *
+     *  Follows POSIX semantics:
+     *
+     *  * if no data is available, and non-blocking set return -EAGAIN
+     *  * if no data is available, and blocking set, wait until data is available
+     *  * If any data is available, call returns immediately
+     *
+     *  @param buffer   The buffer to read in to
+     *  @param length   The number of bytes to read
+     *  @return         The number of bytes read, 0 at end of file, negative error on failure
+     */
+    virtual ssize_t read(void* buffer, size_t length) {
+
+        if(length != sizeof(T)) {
+            return -EINVAL;
+        }
+
+        if(_mail.empty() && !_blocking) {
+            return -EAGAIN;
+        }
+
+        bool mail_was_full = _mail.full();
+
+        osEvent mail_event = _mail.get();
+
+        if(mail_event.status != osEventMail) {
+            return -EIO;
+        }
+
+        T * message = (T *) mail_event.value.p;
+
+        memcpy(buffer, message, sizeof(T));
```

Comment:
Depending on the requirements of `buffer` you should use either placement new with copy construction or copy assignment.